skills/__tests__/hears.test.js

```diff
@@ -0,0 +1,108 @@
+const { helpHandler, extractMessageContents } = require('../hears');
+const { MAIN_HELP_TEXT } = require('../../static');
+
+jest.mock('../../lib/asyncBot');
+
+const TEST_USER_ID = 'TEST_USER';
+const TEST_USER_NAME = 'TEST_USER_NAME';
+const LEARNER_OBJECT = { id: TEST_USER_ID, name: TEST_USER_NAME };
+
+const TEACHER_1_ID = 'TEACHER1';
+const TEACHER_1_NAME = 'TEACHER_1_NAME';
+const TEACHER_2_ID = 'TEACHER2';
+const TEACHER_2_NAME = 'TEACHER_2_NAME';
+
+const mockBot = new (jest.fn())();
+mockBot.whisper = jest.fn();
+
+// These all get passed to asyncBot in extractMessageContents
+mockBot.api = jest.fn();
+mockBot.api.users = jest.fn();
```

Comment:
IMO the `asyncBot` mock should contain the `jest.fn()` implementations.
The tests can then just use `mockBot.api.users.mockImplementation() => { ... })` without needing to do the boilerplate.

Reply:
I'm sure you're aware already but be conscious that the mocks here are defined globally in the file and will apply to everything here with little ability to configure on a per-test basis.

Reply:
I tried, but that led to my not being able to share constants across files (`TEST_USER_ID`, `TEST_USER_NAME`, etc.) which would lead to magic strings. Happy to hear about ways of fixing that, or if you think it's better to just have magic strings anyway.

And yes to the second point, but that didn't seems problematic at this stage as the mock functions fairly closely to how the function actually works. I would consider making it neater as and when it became problematic.

Reply:
if your `asyncBot.js` mock looks like this
```javascript
module.exports = {
  api: {
    users: { 
      info: jest.fn() 
    }
  }
};
```

Then each tests that uses new functions can continue to expand the representation of the mock as required.

All of the above lines go away and you can do this inside each tests that uses that function.

```javascript
mockBot.api.users.info.mockResolvedValue({ user: { name: TEST_USER_NAME } })
```

Since this function is presumably tested by you, or the package maintainer, proving that when passes X returns Y is outside the scope of these tests here.

Reply:
I don't think that works because it's not the `asyncBot` that gets directly used in the `extractMessageContents` function:

```const asyncBot = genAsyncBot(bot);```

The reason I have it set against my `mockBot` in the test is because my mocked `genAsyncBot` just passes back the bot that's given to it.

I _could_ sneak the return value in somewhere in the bot params, but that feels real dirty.

Reply:
Tried a thing anyway, not sure how I feel about it though so happy to shoot in a different direction if you know of a better one.